packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -127,19 +128,23 @@ export class PendingStateManager implements IDisposable {
 			this.initialMessages.isEmpty(),
 			0x2e9 /* "Must call getLocalState() after applying initial states" */,
 		);
-		const newSavedOps = [...this.savedOps].filter(
-			(message) => message.referenceSequenceNumber >= (snapshotSequenceNumber ?? 0),
-		);
+		const newSavedOps = [...this.savedOps].filter((message) => {
+			assert(
+				message.sequenceNumber !== undefined,
+				"saved op should already have a sequence number",
+			);
+			return message.sequenceNumber >= (snapshotSequenceNumber ?? 0);
```

Comment:
@dannimad, I thought a bit about Tony's point "it does not matter", and I want to share some thoughts here on that.
Looking again at existing code, I agree it should not matter, assuming the invariant about when refresh happens holds.
So maybe we are missing clear comment in the code why it's correct.

At the same time, I've shared earlier pointer to PR that shows that some ops are excluded from "dirty" calculation (it's same ID compressor ops :), but the case is broader that just that). From that angle, existing code, before or after this PR is not sufficient. As Loader layer has no clue when its safe to update snapshot. While not ideal, if you broaden the definition of "safe" to be !ConnectionManager.shouldJoinWrite(), then it would be correct, as it includes these cases.
Maybe that's it, but to be honest it's very hard to follow all these implicit contracts in the code. So, I'd rather have more explicit mechanism to ask runtime these things.

The nice thing about explicit ask is that runtime can remember updated snapshot's sequence number, and use it (or assert it's equal to snapshotSequenceNumber), and thus it's easier for readers to follow where it comes from and why it's safe assumption that all ops in pending state will / have to satisfy this requirement.

Reply:
But I think we still have unanswered question - if it does not matter - why the hack in Tyler's test is required? It's not a question of - can test be written without this hack. It's a question - why making this change here changes the need to have this hack? Maybe it's accidental and test issue but would be great to better understand that.

Reply:
I'll work on broadening our safe state in a subsequent PR

Reply:
Also, I'll improve test coverage to see if we hit Tyler's issue or a test shenanigan was the cause of needing such a hack.